File: nff/commands/diagnose.py
"""nff diagnose — classify an ESP32 crash locally (no login, no network, no API key).

The frictionless sibling of `nff repair`: everything runs on this machine using the
rule-based classifier in nff.tools.diagnose. Output is structured facts (crash class,
confidence, extracted registers/backtrace) — symbolized frames and the platform's full
diagnosis remain the `repair` path.
"""

import json as json_module

import click

from nff.tools import diagnose as diagnose_tools
from nff.tools.serial import serial_read, _resolve_port, _resolve_baud, SerialError


@click.command()
@click.option("--serial", "serial_text", default=None, help="Crash/serial output to classify")
@click.option("--capture-ms", default=None, type=int, metavar="MS",
              help="Capture serial from the attached board for N ms, then classify")
@click.option("--port", default=None)
@click.option("--baud", default=None, type=int)
@click.option("--json", "as_json", is_flag=True, help="Print the full structured-facts JSON")
def diagnose(serial_text, capture_ms, port, baud, as_json):
    """Classify a crash from serial output — fully local and offline."""
    if serial_text is None and capture_ms is not None:
        try:
            p = _resolve_port(port)
            b = _resolve_baud(baud)
        except SerialError as exc:
            raise click.ClickException(str(exc))
        serial_text = serial_read(capture_ms, p, b)

    if not serial_text:
        raise click.ClickException("No serial output — pass --serial or --capture-ms")

    result = diagnose_tools.diagnose(serial_text)

    if as_json:
        click.echo(json_module.dumps(result, indent=2))
        return

    if not result["ok"]:
        click.echo(f"No crash found: {result['error']}")
        return

    click.echo(f"{result['title']} ({result['crash_class']}) — "
               f"confidence {result['confidence']:.2f}, family {result['family']}")
    click.echo(f"  {result['rationale']}")
    if result["is_symptom"]:
        click.echo("  note: this class is a SYMPTOM — the root cause is whatever blocked.")
    for candidate in result["candidates"]:
        click.echo(f"  also possible: {candidate['crash_class']} — {candidate['explanation']}")
    click.echo(f"  hint: {result['remediation_hint']}")
    if result["backtrace"]:
        click.echo("  backtrace (unsymbolized): " + " ".join(result["backtrace"]))
        click.echo("  (symbolized frames: `nff repair` with a platform login)")

File: nff-rs/nff/src/commands/diagnose.rs
//! nff diagnose — classify an ESP32 crash locally (no login, no network, no API key).
//!
//! The frictionless sibling of `nff repair`: everything runs on this machine using the
//! rule-based classifier in tools::diagnose. Output is structured facts (crash class,
//! confidence, extracted registers/backtrace) — symbolized frames and the platform's full
//! diagnosis remain the `repair` path.
//!
//! Port of the Python `nff/commands/diagnose.py`.

use anyhow::{anyhow, bail, Result};

use crate::cli::DiagnoseArgs;
use crate::tools::{diagnose as diagnose_tools, serial};

pub fn run(args: &DiagnoseArgs) -> Result<()> {
    let mut serial_text = args.serial.clone();
    if serial_text.is_none() {
        if let Some(capture_ms) = args.capture_ms {
            let port = serial::resolve_port(args.port.as_deref()).map_err(|e| anyhow!("{e}"))?;
            let baud = serial::resolve_baud(args.baud).map_err(|e| anyhow!("{e}"))?;
            serial_text = Some(serial::serial_read(capture_ms, Some(&port), Some(baud)));
        }
    }

    let Some(text) = serial_text.filter(|s| !s.is_empty()) else {
        bail!("No serial output — pass --serial or --capture-ms");
    };

    let result = diagnose_tools::diagnose(&text);

    if args.json {
        println!("{}", serde_json::to_string_pretty(&result)?);
        return Ok(());
    }

    if result["ok"] != true {
        println!(
            "No crash found: {}",
            result["error"].as_str().unwrap_or_default()
        );
        return Ok(());
    }

    let text_of = |key: &str| result[key].as_str().unwrap_or_default().to_string();
    println!(
        "{} ({}) — confidence {:.2}, family {}",
        text_of("title"),
        text_of("crash_class"),
        result["confidence"].as_f64().unwrap_or(0.0),
        text_of("family")
    );
    println!("  {}", text_of("rationale"));
    if result["is_symptom"] == true {
        println!("  note: this class is a SYMPTOM — the root cause is whatever blocked.");
    }
    if let Some(candidates) = result["candidates"].as_array() {
        for candidate in candidates {
            println!(
                "  also possible: {} — {}",
                candidate["crash_class"].as_str().unwrap_or_default(),
                candidate["explanation"].as_str().unwrap_or_default()
            );
        }
    }
    println!("  hint: {}", text_of("remediation_hint"));
    let backtrace: Vec<&str> = result["backtrace"]
        .as_array()
        .map(|frames| frames.iter().filter_map(|f| f.as_str()).collect())
        .unwrap_or_default();
    if !backtrace.is_empty() {
        println!("  backtrace (unsymbolized): {}", backtrace.join(" "));
        println!("  (symbolized frames: `nff repair` with a platform login)");
    }
    Ok(())
}
